File: frontend/src/types/owner.ts
import type { Pet } from "./pet";
import type { Invoice } from "./invoice";

export type Owner = {
  id: number;
  firstName: string;
  lastName: string;
  phone: string;
  email: string;
  address: string;
  petCount: number;
  archived: boolean;
  createdAt: string;
  updatedAt: string;
};

export type CreateOwnerRequest = {
  firstName: string;
  lastName: string;
  phone: string;
  email: string;
  address: string;
};

export type UpdateOwnerRequest = CreateOwnerRequest;

export type OwnerDetail = {
  id: number;
  firstName: string;
  lastName: string;
  phone: string;
  email: string;
  address: string;
  petCount: number;
  pets: Pet[];
  invoices: Invoice[];
  archived: boolean;
  createdAt: string;
  updatedAt: string;
};

export type OwnerStatsResponse = {
  totalOwners: number;
  totalPets: number;
  newOwnersThisMonth: number;
};

File: backend/src/main/java/com/efe/veterinaryclinic/owner/Owner.java
package com.efe.veterinaryclinic.owner;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.Getter;

import java.time.LocalDateTime;

@Entity
@Table(name = "owners")
@Getter
public class Owner {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100)
    private String firstName;

    @Column(nullable = false, length = 100)
    private String lastName;

    @Column(nullable = false, length = 20)
    private String phone;

    @Column(nullable = false, length = 254)
    private String email;

    @Column(nullable = false, length = 500)
    private String address;

    @Column(nullable = false)
    private boolean archived;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    protected Owner() {
    }

    public Owner(String firstName, String lastName, String phone, String email, String address) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.phone = phone;
        this.email = email;
        this.address = address;
        this.archived = false;
    }

    public void update(String firstName, String lastName, String phone, String email, String address) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.phone = phone;
        this.email = email;
        this.address = address;
    }

    public void archive() {
        this.archived = true;
    }

    public void activate() {
        this.archived = false;
    }

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = this.createdAt;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}
